File: Progress-Sheet/0x10-Sorting-II/100-mergesort.py
def merge(left_half,right_half) -> List[int]:
    merged = []
    i = j = 0
    
    while i < len(left_half) and j < len(right_half):
        if left_half[i] <= right_half[j]:
            merged.append(left_half[i])
            i += 1
        else:
            merged.append(right_half[j])
            j += 1
    merged.extend(left_half[i:])
    merged.extend(right_half[j:])
    return merged

def mergeSort(left, right, arr):
    if left == right:
        return [arr[left]]
    mid = left + (right - left) // 2
    left_half = mergeSort(left, mid, arr)
    right_half = mergeSort(mid + 1, right, arr)
   
    return merge(left_half, right_half)

def test():
    assert mergeSort(0, 5, [3, 0, 2, -5, 10, 2]) == [-5, 0, 2, 2, 3, 10], "Not Implemented Properly"
    assert mergeSort(0, 2, [1, 2, 3]) == [1, 2, 3], "Not Implemented Properly"
    assert mergeSort(0, 2, [3, 2, 1]) == [1, 2, 3], "Not Implemented Properly"
    print("Great Job !!!")
test()

File: Progress-Sheet/0x10-Sorting-II/100-mergesort.cpp
#include <bits/stdc++.h>
#define pb push_back
using namespace std;

void merge(vector<int> &a,int low,int mid , int high){
    
    int left = mid;
    int right = mid + 1;
    vector<int> temp; 
    while (left <= mid && right  <= high){
        if (a[left] <= a[right]){
            temp.pb(a[left]);
        }
        else{
            temp.pb(a[right]);
        }

        while (left <= mid){
            temp.pb(a[left]);
        }
        while( right <= mid){
            temp.pb(a[right]);
        }
    }

}

void mS(vector<int> &a , int low, int high){
    if (low >= high) return;
    
    int mid = (low + high) >> 2;

    mS(a,low,mid);
    mS(a,low,mid);
    merge(a,low,mid,high);


}

int main()
{
    int n = 10;
    vector<int> arr(n);
    arr = {1,9,8,7,6,4,2,3,5,10};
    mS(arr,0,n-1);

}